### DISCUSSION TITLE
Invoking onError callback from useQuery

### DISCUSSION THREAD

**Discussion by andy4Expedia** [TanStack/query#6947] [Q&A] [answered]
### Hi team

I know this issue has been answered in varying ways but just as the answer presents itself the issue is closed so I am hoping to get an answer related to title. 

I am trying to unit test my QueryClientProvider and provided what that looks like below.
```
const queryClientConfiguration = (errorComp: ErrorComponent): QueryClient =>
  new QueryClient({
    queryCache: new QueryCache({
      onError: (error): void => {
       errorComp.pushErrorMessages('global', 'response-error', (error as Error).message);
      }
    }),
    mutationCache: new MutationCache({
      onError: (error): void => {
        const err = JSON.parse((error as Error).message);
        const errorMessages = pushErrorMessage(err.statusCode);
        errorComp.pushErrorMessages(errorMessages.key, errorMessages.id, errorMessages.message);
      }
    }),
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
        retry: false,
      }
    }
  });

export const CustomQueryClientProvider: React.FC = ({children}) => {
    return(
        <QueryClientProvider client={queryClientConfiguration(globalErrorStore)}>
            {children}
        </QueryClientProvider>
  )};
```

From the documentation I know the onError call back is invoked with a Rejected promise so i created a custom hook for that. 

```
const useCustomHook = (key: QueryKey): UseQueryResult => useQuery(key, async () => jest.fn().mockRejectedValue(new Error('Error')));

```

The issue I am facing is there doesn't seem to be any change to the query cache itself and any error functions return false. I am looking for a way to invoke that onError call back to trigger the global banner component within the query cache (and subsequently mutation cache when it comes to it)

Any help is appreciated.

**Comment by TkDodo** [comment, marked as answer]
1) are you aware that your query client is not stable? https://tkdodo.eu/blog/react-query-fa-qs#2-the-queryclient-is-not-stable
2) are you sure your `queryFn` returns a rejected Promise? Why not just do: `() => Promise.reject(new Error("Error"))
3) in tests, you pretty likely want to turn off retries or it will time out: https://tkdodo.eu/blog/testing-react-query#turn-off-retries
